lib/Dialect/Pipeline/Transforms/ExplicitRegs.cpp

```diff
@@ -0,0 +1,163 @@
+//===- ExplicitRegs.cpp - Explicit regs pass --------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions of the explicit regs pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Support/BackedgeBuilder.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace pipeline;
+
+namespace {
+
+class ExplicitRegsPass : public ExplicitRegsBase<ExplicitRegsPass> {
+public:
+  void runOnOperation() override;
+
+private:
+  // Recursively routes value v backwards through the pipeline, adding new
+  // registers to 'stage' if the value was not already registered in the stage.
+  // Returns the registerred version of 'v' through 'stage'.
+  Value routeThroughStage(OpOperand &v, PipelineStageOp stage);
+
+  // A mapping storing whether a given stage register constains a registerred
+  // version of a given value. The registered version will be a backedge during
+  // pipeline body analysis. Once the entire body has been analyzed, the
+  // pipeline.stage operations will be replaced with pipeline.stage.register
+  // operations containing the requested regs, and the backedge will be
+  // replaced. MapVector ensures deterministic iteration order, which in turn
+  // ensures determinism during stage op IR emission.
+  DenseMap<PipelineStageOp, llvm::MapVector<Value, Backedge>> stageRegMap;
+
+  // A linked list of stages in the pipeline. Allows for easily looking up the
+  // predecessor stage of a given stage.
+  DenseMap<PipelineStageOp, PipelineStageOp> stagePredecessor;
+
+  std::shared_ptr<BackedgeBuilder> bb;
+};
+
+} // end anonymous namespace
+
+// NOLINTNEXTLINE(misc-no-recursion)
+Value ExplicitRegsPass::routeThroughStage(OpOperand &v, PipelineStageOp stage) {
+  if (!stage) {
+    // Recursive base case - nothing to route (v is a block operand).
+    return v.get();
+  }
+
+  auto regIt = stageRegMap[stage].find(v.get());
+  if (regIt != stageRegMap[stage].end()) {
+    // 'v' is already registered in 'stage'.
+    return regIt->second;
+  }
+
+  auto *definingOp = v.get().getDefiningOp();
+  auto blockPos = [&](Operation *op) {
+    return std::distance(op->getBlock()->begin(), op->getIterator());
```

Comment:
I would try to use isBeforeInBlock instead of std::distance directly. isBeforeInBlock jumps through some hoops to try and amortize the lookup to constant time on average.

Reply:
Awesome, wasn't aware of that function - updated!